sdk/python/kfp_tekton/compiler/_data_passing_rewriter.py

```diff
@@ -52,14 +53,615 @@ def fix_big_data_passing(workflow: List[Dict[Text, Any]]) -> List[Dict[Text, Any
+            )
+        for pipelinerun in pipelinerun_templates:
+            pipeline, pipelinerun_workspaces = bigger_data_passing_pipelinerun(
+                pipelinerun, pipeline_workspaces)
+
+    # Use workspaces to tasks if bigger data passing instead of 'results', 'copy-inputs'
+    for task_template in container_templates:
+        task_template = bigger_data_passing_tasks(
+            task_template, inputs_consumed_as_artifacts,
+            outputs_consumed_as_artifacts)
+
+    # Create pvc for pipelinerun if bigger data passing.
+    # As we used workspaces in tekton pipelines which depends on it.
+    # User need to create PV manually, or enable dynamic volume provisioning, refer to the link of:
+    # https://kubernetes.io/docs/concepts/storage/dynamic-provisioning
+    # TODO: Remove PVC if Tekton version > = 0.12, use 'volumeClaimTemplate' instead
+    if pipelinerun_workspaces:
+        for pipelinerun in pipelinerun_workspaces:
+            workflow.append(create_pvc(pipelinerun))
+
+    # Remove input parameters unless they're used downstream.
+    # This also removes unused container template inputs if any.
+    for template in container_templates + pipeline_templates:
+        spec = template.get('spec', {})
+        spec['params'] = [
+            input_parameter for input_parameter in spec.get('params', [])
+            if (template.get('metadata', {}).get('name'),
+                input_parameter['name']) in inputs_consumed_as_parameters
+        ]
+
+    # Remove output parameters unless they're used downstream
+    for template in container_templates + pipeline_templates:
+        spec = template.get('spec', {})
+        spec['results'] = [
+            output_parameter for output_parameter in spec.get('results', [])
+            if (template.get('metadata', {}).get('name'),
+                output_parameter['name']) in outputs_consumed_as_parameters
+        ]
+
+    # Remove pipeline task parameters unless they're used downstream
+    for template in pipeline_templates:
+        tasks = template.get('spec', {}).get('tasks', [])
+        for task in tasks:
+            task['params'] = [
+                parameter_argument
+                for parameter_argument in task.get('params', [])
+                if (task['taskRef']['name'], parameter_argument['name']
+                    ) in inputs_consumed_as_parameters and
+                (task['taskRef']['name'], parameter_argument['name']
+                 ) not in inputs_consumed_as_artifacts
+                or task['taskRef']['name'] in resource_template_names
+            ]
+
+    # Need to confirm:
+    # I didn't find the use cases to support workflow parameter consumed as artifacts downstream in tekton.
+    # Whether this case need to be supporting?
+
+    clean_up_empty_workflow_structures(workflow)
     return workflow
+
+
+# Replaced '_' to '-' to make the format in tekton unified for params, artifacts, outputs.
```

Comment:
we have a general method that ["sanitizes" Kubernetes names](https://github.com/kubeflow/kfp-tekton/blob/2c68de4/sdk/python/kfp_tekton/compiler/_k8s_helper.py#L20) which should be used instead of `replace('_', '-')`

...which should have been invoked already by [`_sanitize_and_inject_artifact()`](https://github.com/kubeflow/kfp-tekton/blob/d806bc8/sdk/python/kfp_tekton/compiler/compiler.py#L534). If some names were missed it should be added there.

Reply:
Hi @ckadner ,
I tried to add the name and use _sanitize_and_inject_artifact() almost whole day yesterday, but failed, the major reasons:
1.  The params and artifacts cannot be converted before workflow and templates created. otherwise, some value will be mis-matched, it will impact to other features, e.g: loop, conditions. etc...
2. _sanitize_and_inject_artifact() is not suitable to use after workflow created, because a. `_sanitize_and_inject_artifact` will remove the symbol of "$,{,}, etc..", b.  `allow_capital_underscore` will convert  capital and underscore in the same time

Personally, I think I can provide another PR separately, try to unified the params in bigger data and _sanitize_and_inject_artifact. How do you think?